File: chapter10/lists_and_recursion_5.ex
defmodule MyEnum do
  def all?(list, fun \\ &(&1))
  def all?([], _), do: true
  def all?([head | tail], fun), do: !!fun.(head) and all?(tail, fun)

  def each([], _), do: :ok
  def each([head | tail], fun) do
    fun.(head)
    each(tail, fun)
  end

  def filter([], _), do: []
  def filter([head | tail], fun) do
    if fun.(head) do
      [head | filter(tail, fun)]
    else
      filter(tail, fun)
    end
  end

  def split([], _), do: {[], []}
  def split(list, count) when count == 0, do: {[], list}
  def split([head | tail], count) when count > 0 do
    {first, rest} = split(tail, count - 1)
    {[head | first], rest}
  end
  def split(list, count) when count < 0, do: split(list, max(0, length(list) + count))

  def take([], _), do: []
  def take(_, amount) when amount == 0, do: []
  def take([head | tail], amount) when amount > 0, do: [head | take(tail, amount - 1)]
  def take(list, amount) when amount < 0,
    do: list |> Enum.reverse |> take(-amount) |> Enum.reverse
end

File: chapter10/lists_and_recursion_5.clj
(ns lists-and-recursion-5
  (:refer-clojure :exclude [every? filter run! split-at take]))

(defn every? [pred coll]
  (if (empty? coll)
    true
    (and (boolean (pred (first coll)))
         (recur pred (rest coll)))))

(defn run! [proc coll]
  (when (seq coll)
    (proc (first coll))
    (recur proc (rest coll))))

(defn filter [pred coll]
  (lazy-seq
   (when-let [[head & tail] (seq coll)]
     (if (pred head)
       (cons head (filter pred tail))
       (filter pred tail)))))

(defn split-at [n [head & tail :as coll]]
  (cond
    (empty? coll) [() ()]
    (not (pos? n)) [() (seq coll)]
    :else (let [[fst snd] (split-at (dec n) tail)]
            [(cons head fst) snd])))

(defn take [n coll]
  (lazy-seq
   (when (and (pos? n) (seq coll))
     (cons (first coll) (take (dec n) (rest coll))))))
